File: src/backend-rust/src/routes/recovery.rs
use std::collections::HashMap;

use axum::{
    Json, Router,
    extract::{Extension, Path, Query, State},
    routing::get,
};
use paladinscat_core::{
    database::Database,
    web_compat::{paginate, parse_js_integer},
};
use serde_json::{Value, json};
use time::{OffsetDateTime, format_description::well_known::Rfc3339};

use crate::{error::ApiError, request::RequestId};

#[derive(Clone)]
struct RecoveryState {
    database: Database,
}

pub fn router(database: Database) -> Router {
    Router::new()
        .route("/recovery/broken-skins", get(broken_skins))
        .route(
            "/recovery/broken-skins/{champion_id}",
            get(broken_skins_for_champion),
        )
        .route("/recovery/stats", get(recovery_stats))
        .route("/recovery/stats/{match_id}", get(recovery_stats_for_match))
        .route("/recovery/pending", get(pending_recovery))
        .with_state(RecoveryState { database })
}

async fn broken_skins(
    State(state): State<RecoveryState>,
    Extension(request_id): Extension<RequestId>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<Vec<Value>>, ApiError> {
    let champion_id = match query.get("championId").filter(|value| !value.is_empty()) {
        Some(value) => {
            Some(parse_js_integer(value).ok_or_else(|| ApiError::internal(&request_id))?)
        }
        None => None,
    };
    let rows = if let Some(champion_id) = champion_id {
        let champion_id =
            i32::try_from(champion_id).map_err(|_| ApiError::internal(&request_id))?;
        state
            .database
            .query_json(
                "SELECT * FROM broken_skins WHERE champion_id = $1 ORDER BY champion_id, skin_id",
                &[&champion_id],
            )
            .await
            .map_err(|error| ApiError::database(error, &request_id))?
    } else {
        state
            .database
            .query_json(
                "SELECT * FROM broken_skins ORDER BY champion_id, skin_id",
                &[],
            )
            .await
            .map_err(|error| ApiError::database(error, &request_id))?
    };
    Ok(Json(rows))
}

async fn broken_skins_for_champion(
    State(state): State<RecoveryState>,
    Extension(request_id): Extension<RequestId>,
    Path(champion_id): Path<String>,
) -> Result<Json<Vec<Value>>, ApiError> {
    let champion_id = parse_js_integer(&champion_id)
        .ok_or_else(|| ApiError::validation("Invalid champion ID"))
        .and_then(|value| i32::try_from(value).map_err(|_| ApiError::internal(&request_id)))?;
    let rows = state
        .database
        .query_json(
            "SELECT * FROM broken_skins WHERE champion_id = $1 ORDER BY skin_id",
            &[&champion_id],
        )
        .await
        .map_err(|error| ApiError::database(error, &request_id))?;
    Ok(Json(rows))
}

async fn recovery_stats(
    State(state): State<RecoveryState>,
    Extension(request_id): Extension<RequestId>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<Vec<Value>>, ApiError> {
    let pagination = paginate(
        query.get("page").map(String::as_str),
        query.get("perPage").map(String::as_str),
    );
    let from = query.get("from").filter(|value| !value.is_empty());
    let to = query.get("to").filter(|value| !value.is_empty());
    let rows = match (from, to) {
        (Some(from), Some(to)) => {
            let from = parse_javascript_date(from, &request_id)?;
            let to = parse_javascript_date(to, &request_id)?;
            state
                .database
                .query_json(
                    "SELECT * FROM recovery_stats WHERE created_at >= $1 AND created_at <= $2 ORDER BY created_at DESC LIMIT $3 OFFSET $4",
                    &[&from, &to, &pagination.per_page, &pagination.offset],
                )
                .await
                .map_err(|error| ApiError::database(error, &request_id))?
        }
        (Some(from), None) => {
            let from = parse_javascript_date(from, &request_id)?;
            state
                .database
                .query_json(
                    "SELECT * FROM recovery_stats WHERE created_at >= $1 ORDER BY created_at DESC LIMIT $2 OFFSET $3",
                    &[&from, &pagination.per_page, &pagination.offset],
                )
                .await
                .map_err(|error| ApiError::database(error, &request_id))?
        }
        (None, Some(to)) => {
            let to = parse_javascript_date(to, &request_id)?;
            state
                .database
                .query_json(
                    "SELECT * FROM recovery_stats WHERE created_at <= $1 ORDER BY created_at DESC LIMIT $2 OFFSET $3",
                    &[&to, &pagination.per_page, &pagination.offset],
                )
                .await
                .map_err(|error| ApiError::database(error, &request_id))?
        }
        (None, None) => state
            .database
            .query_json(
                "SELECT * FROM recovery_stats ORDER BY created_at DESC LIMIT $1 OFFSET $2",
                &[&pagination.per_page, &pagination.offset],
            )
            .await
            .map_err(|error| ApiError::database(error, &request_id))?,
    };
    Ok(Json(rows))
}

fn parse_javascript_date(value: &str, request_id: &RequestId) -> Result<OffsetDateTime, ApiError> {
    if value.len() == 10 {
        let midnight = format!("{value}T00:00:00.000Z");
        if let Ok(parsed) = OffsetDateTime::parse(&midnight, &Rfc3339) {
            return Ok(parsed);
        }
    }
    OffsetDateTime::parse(value, &Rfc3339).map_err(|error| {
        tracing::error!(value, error = %error, "date query parameter could not be parsed");
        ApiError::internal(request_id)
    })
}

async fn recovery_stats_for_match(
    State(state): State<RecoveryState>,
    Extension(request_id): Extension<RequestId>,
    Path(match_id): Path<String>,
) -> Result<Json<Value>, ApiError> {
    let match_id =
        parse_js_integer(&match_id).ok_or_else(|| ApiError::validation("Invalid match ID"))?;
    let row = state
        .database
        .one_json(
            "SELECT * FROM recovery_stats WHERE match_id = $1",
            &[&match_id],
        )
        .await
        .map_err(|error| ApiError::database(error, &request_id))?
        .ok_or_else(|| {
            ApiError::not_found("Recovery stats not found", json!({ "matchId": match_id }))
        })?;
    Ok(Json(row))
}

async fn pending_recovery(
    State(state): State<RecoveryState>,
    Extension(request_id): Extension<RequestId>,
    Query(query): Query<HashMap<String, String>>,
) -> Result<Json<Vec<Value>>, ApiError> {
    let parsed_limit = query.get("limit").and_then(|value| parse_js_integer(value));
    let limit = match parsed_limit {
        None | Some(0) => 50,
        Some(limit) => limit.min(200),
    };
    let rows = state
        .database
        .query_json(
            "SELECT entity_id, entity_type, status, endpoint, created_at FROM raw_ingest_buffer WHERE entity_type = 'match' AND status IN ('pending', 'failed') ORDER BY created_at DESC LIMIT $1",
            &[&limit],
        )
        .await
        .map_err(|error| ApiError::database(error, &request_id))?;
    Ok(Json(rows))
}

File: legacy/src-backend/routes/recovery.ts
import { FastifyInstance } from 'fastify';
import { query, one } from '../config/db';
import { paginate } from '../utils/query-helpers';
import { FilterBuilder } from '../utils/filter-builder';

/**
 * Recovery Diagnostics Routes.
 * Exposes broken skin registry and recovery audit trail.
 * Tables: broken_skins, recovery_stats, raw_ingest_buffer.
 */
export default async function recoveryRoutes(fastify: FastifyInstance) {
  /**
   * GET /recovery/broken-skins — List known broken skin IDs
   * ?championId= filter
   */
  fastify.get('/broken-skins', async (req: any) => {
    const fb = new FilterBuilder();
    if (req.query.championId) fb.eq('champion_id', parseInt(req.query.championId, 10));

    const { clause, params } = fb.build();
    const rows = await query(`SELECT * FROM broken_skins${clause} ORDER BY champion_id, skin_id`, params);
    return rows;
  });

  /**
   * GET /recovery/broken-skins/:championId — Broken skins for a specific champion
   */
  fastify.get('/broken-skins/:championId', async (req: any, reply: any) => {
    const championId = parseInt(req.params.championId, 10);
    if (!Number.isInteger(championId)) {
      return reply.status(400).send({ error: { code: 'VALIDATION', message: 'Invalid champion ID' } });
    }

    const rows = await query('SELECT * FROM broken_skins WHERE champion_id = $1 ORDER BY skin_id', [championId]);
    return rows;
  });

  /**
   * GET /recovery/stats — Recovery stats overview
   * ?from=, ?to=, ?page=, ?perPage=
   */
  fastify.get('/stats', async (req: any) => {
    const { page, perPage, offset } = paginate({ page: req.query.page, perPage: req.query.perPage });

    const fb = new FilterBuilder();
    if (req.query.from) fb.gte('created_at', new Date(req.query.from));
    if (req.query.to) fb.lte('created_at', new Date(req.query.to));

    const { clause, params } = fb.build();
    const rows = await query(
      `SELECT * FROM recovery_stats${clause} ORDER BY created_at DESC LIMIT $${params.length + 1} OFFSET $${params.length + 2}`,
      [...params, perPage, offset]
    );
    return rows;
  });

  /**
   * GET /recovery/stats/:matchId — Recovery details for a specific match
   */
  fastify.get('/stats/:matchId', async (req: any, reply: any) => {
    const matchId = parseInt(req.params.matchId, 10);
    if (!Number.isInteger(matchId)) {
      return reply.status(400).send({ error: { code: 'VALIDATION', message: 'Invalid match ID' } });
    }

    const row = await one('SELECT * FROM recovery_stats WHERE match_id = $1', [matchId]);
    if (!row) {
      return reply.status(404).send({ error: { code: 'NOT_FOUND', message: 'Recovery stats not found', details: { matchId } } });
    }
    return row;
  });

  /**
   * GET /recovery/pending — Matches pending recovery (from raw_ingest_buffer)
   * Shows entries with status='pending' or status='failed' for match entities
   * ?limit=
   */
  fastify.get('/pending', async (req: any) => {
    const limit = Math.min(parseInt(req.query.limit as string) || 50, 200);

    const rows = await query(
      `SELECT entity_id, entity_type, status, endpoint, created_at
       FROM raw_ingest_buffer
       WHERE entity_type = 'match' AND status IN ('pending', 'failed')
       ORDER BY created_at DESC
       LIMIT $1`,
      [limit]
    );
    return rows;
  });
}
